### ISSUE TITLE
Looking for maintainer(s)

### ISSUE THREAD

**Issue by isagalaev** [highlightjs/highlight.js#1678]
The project was dormant for a while because none of the current nominal maintainers has enough time/motivation to do the job. Hence, the project is looking for a new maintainer (or several).

## Involvement

Ideally I would gladly hand the project over completely to a new principal maintainer, provided they're interested in keeping it roughly within the same principles it's been following the [past 10 years][10]. It would be a real letdown for existing users if the project suddenly pivoted into something else.

Realistically though, since I am probably still the only person intimately familiar with the core engine, we could probably make it work with shared maintenance. We can work out the details down the road.

My ultimate goal here is to stop being a bottleneck.

## Revival plan

Another reason why the project got hard to maintain is that it's outgrown a few old assumption and is in need of reworking to bring it back into maintainable state. Roughly:

- Our build process with several build targets and customizable sets of languages is a pain for both maintainers and users. Since we now live in the world with HTTP/2 and evolved JavaScript packaging it needs to be replaced. I can be instrumental in describing the pain points and historical perspective, but I <s>don't know shit</s> not familiar with modern JS infrastructure.

- It might make sense to severely limit the set of languages supported within the main repository, with others moving outside for community support, hosted somewhere else. For that the API between the library and languages needs to (finally) be stabilized.

## Migration

- We should migrate the project under the org `https://github.com/highlightjs/`, forking it from here. I don't have a good idea what to do with all the open issues on this repo under `/isagalaev/`. One way could be defaulting on all of them here and asking interested people to reopen them at the new location.

- I own highlightjs.org and happy to keep it running for now, but if we decide on handing over maintenance completely, the domain should probably go to the new maintainer as well.

- There's also a Twitter account `@highlightjs` which I don't want to care about anymore and a Google group which might be simply deleted due to lack of traffic/purpose.

## Anyone?

A few people have already pinged me over email, so please reiterate your intention here, to which degree you're interested in helping out/taking over.

And thank you!

[10]: http://softwaremaniacs.org/blog/2016/08/16/highlight-js-turns-10/

**Comment by SirCmpwn** [comment]
Hey @isagalaev, glad you're taking this step for the health of the project. I would be happy to volunteer. I do not have time to work on extended development on the project, but I do have time for maintainership - holding the vision in my head and applying to reviewing pull requests, providing feedback on them, and merging and releasing them.

I agree with your [driving principles](http://softwaremaniacs.org/blog/2016/08/16/highlight-js-turns-10/) and I will happily use them as a guide for my work here.

>Our build process with several build targets and customizable sets of languages is a pain for both maintainers and users. Since we now live in the world with HTTP/2 and evolved JavaScript packaging it needs to be replaced. I can be instrumental in describing the pain points and historical perspective, but I ~~don't know shit~~ not familiar with modern JS infrastructure.

This is a good idea, and shouldn't be terribly difficult.

>We should migrate the project under the org `https://github.com/highlightjs/`, forking it from here. I don't have a good idea what to do with all the open issues on this repo under `/isagalaev/`. One way could be defaulting on all of them here and asking interested people to reopen them at the new location.

You can move the repo to a new org without making a fork, keeping all of the issues and pull requests. I agree that an org is the right home for highlight.js given the refactoring plans. These repos:

>It might make sense to severely limit the set of languages supported within the main repository, with others moving outside for community support, hosted somewhere else. For that the API between the library and languages needs to (finally) be stabilized.

Could live there.

>There's also a Twitter account `@highlightjs` which I don't want to care about anymore and a Google group which might be simply deleted due to lack of traffic/purpose.

FWIW I don't have any interest in maintaining either of these.

**Comment by ldct** [comment]
Hi @isagalaev, as mentioned over email, I'd be happy to help maintain this. The 2 priorities listed in the revival plan make sense to me and would be what I'd focus on doing.

**Comment by duncdrum** [comment]
@isagalaev I m in the same boat as SirCmpwn. 
Moving the repo to highlight.js org here on github shouldn't be a problem. 
As for trimming rare languages there are already a number of good libraries with such a limited support. "Community support" looks great on paper, but in practice I fear that such an outsourcing of languages will lessen the appeal of highlight.js and create more maintenance hassles than its worth.

**Comment by idleberg** [comment]
I'd (still) like to see an officially maintained package on npm, probably using a organization account. Single languages could be published as single packages using the same organization.

**Example:**

```sh
# Install typical web languages
npm i @hljs/core @hljs/html @hljs/css  @hljs/js
```

Since I'm the author of the `nsis.js` and maintainer of several other NSIS language packs, I'd be happy to continue as official maintainer for that language.

**Comment by shiya** [comment]
I'd really like to help because this is such a good library and I've been using it at several places. I have some time to spare but not sure where to get started.

**Comment by sangdth** [comment]
I'm not looking to take over, but I would like to help.

**Comment by DuncanmaMSFT** [comment]
I'm interested in helping, I'd even be up for handling the twitter account :) (I ran the one for channel9.msdn.com for 5 or 6 years)

**Comment by isagalaev** [comment]
Just to give you guys a status update: I'm sitting silent simply to collect all the shout outs for now. I think I'll move the repo and flesh out a few tasks this weekend, so stay tuned!

And thanks a lot, everyone!

**Comment by tracker1** [comment]
I'd be happy to participate in the new org... though not sure how much I would contribute beyond reviewing and merging/publishing pull requests.  I don't have enough time to participate in the refactoring for separating languages, though would really like to be.

**Comment by PatrickMassot** [comment]
Are there any news on this issue? My real question is: how plausible it is that the queue of PR adding new languages will be flushed soonish? Note that I'm not complaining about anything, I only try to get some insight.

**Comment by SirCmpwn** [comment]
I wonder if the best course of action would be to establish a GitHub org and grant all of the volunteers the rights necessary to execute maintainership duties? There needn't necessarily be a single maintainer - more would make the workload easier to handle.

**Comment by isagalaev** [comment]
> I wonder if the best course of action would be to establish a GitHub org and grant all of the volunteers the rights necessary to execute maintainership duties?

Yes, that's the plan. I haven't yet gotten to it because I wanted to write a minimal maintainer's guide with dos and don'ts accumulated over the years. Without one I am bound to remain a bottleneck, this time as a reviewer.

Please be patient! :-)

**Comment by Tyriar** [comment]
+@mjbvz

**Comment by gusbemacbe** [comment]
@Tyriar, has @mjbvz spoken with @isagalaev ?

**Comment by mjbvz** [comment]
Sorry but this is not something I'm interested in. Our usage in highlight js is quite limited. We only use it in the markdown preview, and use it there mainly because many websites also use it for syntax coloring

**Comment by SirCmpwn** [comment]
>Yes, that's the plan. I haven't yet gotten to it because I wanted to write a minimal maintainer's guide with dos and don'ts accumulated over the years. Without one I am bound to remain a bottleneck, this time as a reviewer.

I feel like we might have figured it out by now

**Comment by isagalaev** [comment]
True, and sorry for letting everyone down. Let me find some free evening this week (Wednesday most likely) to just move it over and unblock things.

**Comment by SirCmpwn** [comment]
Thank you!

**Comment by isagalaev** [comment]
OK, let me try this again…

I moved the project over to the new place https://github.com/highlightjs/highlight.js and invited @SirCmpwn as a new maintainer (since he prompted me to finally do it, so he's clearly still interested — thank you!).  I kindly ask all the others who volunteered (@ldct, @idleberg, @shiya, @sangdth, @DuncanmaMSFT, @tracker1) to shout out if they're still interested, it's been half a year ago after all…

I put a very basic [maintainer's guide](https://github.com/highlightjs/highlight.js/blob/master/docs/maintainers-guide.rst) in place, it'll probably grow to include something about governance. We didn't need it before, but will clearly need from now on.

Finally, I created a ticket for a new build system: #1759. I wouldn't go as far as to say that it's essential for the future maintenance of the project, but it's certainly something which only grows in importance with time.

**Comment by sangdth** [comment]
@isagalaev I'm still interested.

**Comment by idleberg** [comment]
@isagalaev I'm still interested.

**Comment by SirCmpwn** [comment]
Thanks @isagalaev! I invited @sangdth and @idleberg as well.

Let me restate my original offer:

>I would be happy to volunteer. I do not have time to work on extended development on the project, but I do have time for maintainership - holding the vision in my head and applying to reviewing pull requests, providing feedback on them, and merging and releasing them.

I still don't have time to do extended development on the project, but I'm going to turn my focus towards the huge backlog of pull requests (122!) we have to get through.

**Comment by ldct** [comment]
@isagalaev I'm still interested

**Comment by marcoscaceres** [comment]
I’m also available for code review.

**Comment by isagalaev** [comment]
@marcoscaceres I see someone added you to the org already, welcome aboard :-) Feels really good to have people doing things that would otherwise be blocked on me!

**Comment by marcoscaceres** [comment]
@isagalaev, could you kindly add @gmmorris to the contributors? Also, could you give "marcosc" the ability publish on npm?

**Comment by marcoscaceres** [comment]
Hi folks who offered to volunteer. Could I get a show of hands from those of you who are still willing to help here?

**Comment by marcoscaceres** [comment]
@gusbemacbe, January is totally ok :)

**Comment by idleberg** [comment]
Is there any place to discuss, other than this “issue”? I was thinking of Gitter, Slack, Rocket Chat… you name it.

**Comment by marcoscaceres** [comment]
Ok, I set up a slack workspace:
https://highlightjs.slack.com

But I'll need to send individual invite - folks, please email me at mcaceres@mozilla.com so I can add you.

**Comment by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll** [comment]
@marcoscaceres have you thought about using gitter instead? one advantage would be that the discussions would be publicly readable without registration

**Comment by marcoscaceres** [comment]
@IIIIllllIIIIllllIIIIllllIIIIllllIIIIll I have, but it's Slack is fine... we just need to redirect things back to bugs on Github to keep a record of things. Pull requests and bug details serve as better records for technical discussion, while Slack is good for planning and community building.

**Comment by isagalaev** [comment]
Hey guys! I just noticed this thread accidentally, and while I'm mostly away from the project, I'm going to set up my notifications in a way so that I will actually pay attention to direct messages. So if you need me to update highlightjs.org in any way, please tell me!

**Comment by marcoscaceres** [comment]
@isagalaev, much appreciated.

**Comment by isagalaev** [comment]
> @isagalaev, could you kindly add @gmmorris to the contributors?

Done. I also fixed your role to owner, so you can do that too. I was fairly sure everyone I added was oener already, but apparently I missed something.

> Also, could you give "marcosc" the ability publish on npm?

This is supposed to happen automatically when a new release is created (https://github.com/highlightjs/highlight.js/blob/master/docs/maintainers-guide.rst#release-process). And I think it makes sense to keep it this way so all updates happen everywhere simultaneously.

**Comment by marcoscaceres** [comment]
Ok, cool. I'd missed that bit of the maintainers guide.

**Comment by yyyc514** [comment]
No interest in being "principal" maintainer, but I don't mind helping out from time to time with PRs and issues as I can.

**Comment by yyyc514** [comment]
And just FYI: if becoming a maintainer is "scary" for anyone who's maybe never done that before or a little commitment shy... then you don't have to step up and say "I'll be a maintainer" or make a huge commitment...   There is no official swearing in process required to make a pull request.

There are plenty of PRs tagged with `help welcome` or `beginner`... if all you do is drop by and contribute a single PR or two, that would still be greatly appreciated... and maybe you end up staying longer and contributing more, maybe not.

Personally I started with one PR, then another, then 20 more, and then eventually it made sense to add myself as a listed maintainer in `AUTHORS`.

Just wanted to encourage others: you can start small if you want and see where it takes you. :-)

**Comment by NotMoni** [comment]
If it's still possible. I am ready to help and maintain. 🦄

**Comment by yyyc514** [comment]
Cool.  You could start by cutting your teeth on some of the beginning tagged issues:

- https://github.com/highlightjs/highlight.js/issues/2217
- https://github.com/highlightjs/highlight.js/issues/1616
- https://github.com/highlightjs/highlight.js/issues/1559

Or if there is anything specific you're interested in, let me know. The huge CSS consolidation work could use some time invested.

**Comment by NotMoni** [comment]
Sure mate

**Comment by yyyc514** [comment]
Any objections to posting on https://adoptoposs.org ?  I just saw this.

**Comment by egor-rogov** [comment]
> Any objections to posting on https://adoptoposs.org ? I just saw this.

I wouldn't expect it to help find enthusiasts, but why not?

**Comment by allejo** [comment]
> Any objections to posting on [adoptoposs.org](https://adoptoposs.org) ? I just saw this.

No objections here. I hope something comes out of this :sweat_smile:

**Comment by bradleymackey** [comment]
Happy to help with some basic PR maintenance: approving PRs runs, closing PRs for new core grammars (with suggestion to add to the third party list) and validating/closing issues.

**Comment by isagalaev** [comment]
> Hi, seeing that this issue is still open are you still looking for maintainers/contributors/code reviewers?

Hey! I'm not in a position to answer your question myself, but last I know the discussion was now happening in Discord https://discord.gg/M24EbU7ja9, so you may have more luck asking over there.

**Comment by joshgoebel** [comment]
@paladox 

> are you able to add more people as owners/being able to control this repo

This issue #1678 is open and has never closed - we need help and have always needed help - even when I spent a lot of time on this every week.  But we're not going to give those rights to just anyone.  If someone else wants to show serious long-term interest in the project I imagine it would look something like what I did to get here:

- Start answering issues as if you already own the place (if you can help, just help)
- Open PRs for issues you're able to solve (esp ones no one else wants to fix)
- Triage existing PRs and help out with moving them forward - and defer to me when it's unclear how to proceed
  - A lot of it is just maintaining our policies, code styles, making sure people write tests, don't break things or under-think, etc...
- Work with me to get those PRs merged.
- Start being so helpful that it eventually becomes terribly annoying for me to constantly approve work, and that's how you get commit bits. (or at least how I got mine)

Yes, I've been away a while, but I'm back (sorta) now...  and while I don't have the time I had before I think I do have the time to work with anyone who'd be more interested in getting more involved with an eye toward joining the core team long term.  I did just close/merge about 40 PRs and respond to a bunch of issues just today.

CC @allejo

**Comment by paladox** [comment]
Thank you!!

**Comment by Baltsat** [comment]
Hi @joshgoebel — I would like to take you up on this. I have started by landing #4453 and #4454, with #4455 pending. I can commit to a sustained maintenance lane: triaging grammar and parser issues, reproducing abandoned bugs, running the full grammar/build checks on community PRs, and helping authors get changes merge-ready while deferring architecture calls to the core team. My goal is to earn core-team trust over time. Is there a current backlog slice or group of PRs you would prefer I own first?

**Comment by tracker1** [comment]
Not sure if you ever got around to updating language plugins/packaging...

I would suggest keeping each language in the repo, as much as makes sense and release it as separate JS/TS modules within the same package... since the act of initial download into a project is generally a server thing... within the project's usage, the downstream project can then simply include/plugin the platform and language modules they want to use at initialization, which will land their downloads how they see fit for either async or packaging.

Just my own thought on this... I know I don't have the time to do it myself, but that's the direction I've been moving for a few of my own packages I'm working on actively.  The server/project download from the repository is less of an issue so it makes sense to just keep bundling in the package, but separate includes.

I would also suggest an interim migration to TypeScript, if anyone is up for that... possibly a 1:1 port with AI assist.  This would help with future improvements as well as allow for direct publishing to JSR in addition to NPM.

**Comment by davidpavlovschi** [comment]
Hi @joshgoebel — adding my name here as well. You merged my #4456 yesterday (C/C++ raw-string delimiter fix, done with the extra C fixture you asked for). I'd be happy to take a regular lane on grammar/parser issue triage and first-pass review of dialect PRs — whatever offloads the most from you. Happy to coordinate with @Baltsat rather than overlap.
